Allow ActiveStorage::Blob#service_url to pass addition options to service.url

huacnlee opened the pull request:
Allow `ActiveStorage::Blob#service_url` to pass addition options to `service.url`.

Because there have some service needs more parameters for file URL:

https://www.alibabacloud.com/help/doc-detail/44687.htm

```rb
class AlicloudOSSService < Service
  def url(key, options = {})
    image_process = options[:oss_process] || "image/resize,w_800"
    "http://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=#{image_process}"
  end
end
```

https://github.com/huacnlee/activestorage-aliyun/blob/master/lib/active_storage/service/aliyun_service.rb#L43

```erb
<%= image_tag @user.avatar.service_url(oss_process: "image/resize,m_fill,h_100,w_100" %>
```

georgeclaghorn commented on activestorage/app/models/active_storage/blob.rb:
```diff
@@ -109,8 +109,14 @@ def text?
+      filename: filename,
+      content_type: content_type
+    )
+    service.url key, options
```
Can we just splat the extra options onto the `url` call?

```ruby
service.url key, expires_in: expires_in, filename: filename, content_type: content_type,
  disposition: forcibly_serve_as_binary? ? :attachment : disposition, **options
```

huacnlee commented:
👌

George Claghorn <notifications@github.com>于2018年2月1日 周四下午9:11写道：

> *@georgeclaghorn* requested changes on this pull request.
> ------------------------------
>
> In activestorage/app/models/active_storage/blob.rb
> <https://github.com/rails/rails/pull/31854#discussion_r165350149>:
>
> > @@ -109,8 +109,14 @@ def text?
>    # with users. Instead, the +service_url+ should only be exposed as a redirect from a stable, possibly authenticated URL.
>    # Hiding the +service_url+ behind a redirect also gives you the power to change services without updating all URLs. And
>    # it allows permanent URLs that redirect to the +service_url+ to be cached in the view.
> -  def service_url(expires_in: service.url_expires_in, disposition: :inline, filename: self.filename)
> -    service.url key, expires_in: expires_in, disposition: forcibly_serve_as_binary? ? :attachment : disposition, filename: filename, content_type: content_type
> +  def service_url(expires_in: service.url_expires_in, disposition: :inline, filename: self.filename, **options)
> +    options.merge!(
> +      expires_in: expires_in,
> +      disposition: forcibly_serve_as_binary? ? :attachment : disposition,
> +      filename: filename,
> +      content_type: content_type
> +    )
> +    service.url key, options
>
> Can we just splat the extra options onto the url call?
>
> service.url key, expires_in: expires_in, filename: filename, content_type: content_type,
>   disposition: forcibly_serve_as_binary? ? :attachment : disposition, **options
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/31854#pullrequestreview-93273587>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAVjivU9gJOe1r0hXQCkAb2S2sNqGz9ks5tQbf6gaJpZM4R1X32>
> .
>
-- 
李华顺 (Jason Lee) <http://huacnlee.com>
----------------------------------------
http://huacnlee.com
http://twitter.com/huacnlee
http://github.com/huacnlee

huacnlee committed:
Allow `ActiveStorage::Blob#service_url` to pass addition options to `service.url`.

Because there have some service needs more parameters for file URL:

https://www.alibabacloud.com/help/doc-detail/44687.htm

```rb
class AliyunService < Service
  def url(key, options = {})
    image_process = options[:oss_process] || "image/resize,w_800"
    "http://image-demo.oss-cn-hangzhou.aliyuncs.com/example.jpg?x-oss-process=#{image_process}"
  end
end
```

Use case:

```erb
<%= image_tag @user.avatar.service_url(oss_process: "image/resize,m_fill,h_100,w_100" %>
```

huacnlee commented:
@georgeclaghorn I have updated
